servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/MulticastPublisher.java

```diff
@@ -15,355 +15,345 @@
+            final Subscriber<? super T>[] newSubs = (Subscriber<? super T>[]) Array.newInstance(Subscriber.class, 1);
+            newSubs[0] = new TerminalSubscriber<>(t);
+            for (;;) {
+                final Subscriber<? super T>[] subs = subscribers;
+                if (newSubscribersUpdater.compareAndSet(this, subs, newSubs)) {
+                    Throwable delayedCause = null;
+                    for (final Subscriber<? super T> subscriber : subs) {
+                        try {
+                            terminator.accept(subscriber, t);
+                        } catch (Throwable cause) {
+                            delayedCause = catchUnexpected(delayedCause, cause);
+                        }
+                    }
+                    if (delayedCause != null) {
+                        throwException(delayedCause);
+                    }
+                    break;
+                }
             }
-            if (sourceRequestedUpdater.compareAndSet(this, sourceRequested, individualSourceRequested)) {
-                subscription.request(individualSourceRequested - sourceRequested);
-                break;
+        }
+
+        private Completable safeTerminalStateReset(@Nullable Throwable t) {
+            try {
+                return terminalResubscribe.apply(t);
+            } catch (Throwable cause) {
+                LOGGER.warn("terminalStateReset {} threw", terminalResubscribe, cause);
             }
+            return Completable.never();
         }
     }
 
-    void cancelIndividualSubscriber(int subscriberIndex) {
-        Subscriber<? super T> subscriber = subscribers.getAndSet(subscriberIndex, noopSubscriber());
-        if (subscriber != NOOP_SUBSCRIBER && notCancelledCountUpdater.decrementAndGet(this) == 0) {
-            subscription.cancel();
+    private static final class MulticastFixedSubscriber<T> extends MulticastLeafSubscriber<T>
+            implements PriorityQueueNode {
+        private final int index;
+        private final MulticastPublisher<T>.State root;
+        private final Subscriber<? super T> subscriber;
+        private final Subscriber<? super T> ctxSubscriber;
+        /**
+         * Protected by {@link MulticastRootSubscriber} lock.
+         */
+        private long priorityQueueValue;
+        /**
+         * Protected by {@link MulticastRootSubscriber} lock.
+         */
+        private int priorityQueueIndex = INDEX_NOT_IN_QUEUE;
+
+        MulticastFixedSubscriber(final MulticastPublisher<T>.State root,
+                                 final Subscriber<? super T> subscriber, final AsyncContextMap contextMap,
+                                 final AsyncContextProvider contextProvider, final int index) {
+            this.root = root;
+            this.index = index;
+            this.subscriber = requireNonNull(subscriber);
+            // TODO(scott): SignalOffloader requires Subscriber methods invoked in order and terminal
```

Comment:
@bondolo - fyi, this relates to your work in offloading. I think this is a safe assumption but plz confirm.